core/src/datayoga_core/blocks/relational/write/block.py

```diff
@@ -22,45 +23,75 @@ class Block(DyBlock, metaclass=ABCMeta):
+            self.foreach = self.properties.get("foreach")
+            tbl = sa.Table(self.table, sa.MetaData(schema=self.schema), autoload_with=engine)
 
-        if self.db_type in (relational_utils.DbType.SQLSERVER, relational_utils.DbType.ORACLE):
-            # MERGE statement requires this
-            self.connection = self.connection.execution_options(autocommit=True)
+            logger.debug(f"Connecting to {self.db_type}")
+            connection = engine.connect()
 
-        if self.opcode_field:
-            self.business_key_columns = [column["column"] for column in write_utils.get_column_mapping(self.keys)]
-            self.mapping_columns = [column["column"] for column in write_utils.get_column_mapping(self.mapping)]
+            self.engine = engine
+            self.connection = connection
+            self.tbl = tbl
+
+            if self.opcode_field:
+                self.business_key_columns = [column["column"] for column in write_utils.get_column_mapping(self.keys)]
+                self.mapping_columns = [column["column"] for column in write_utils.get_column_mapping(self.mapping)]
+
+                self.columns = self.business_key_columns + [x for x in self.mapping_columns
+                                                            if x not in self.business_key_columns]
 
-            self.columns = self.business_key_columns + [x for x in self.mapping_columns
-                                                        if x not in self.business_key_columns]
+                for column in self.columns:
+                    if not column in tbl.columns:
+                        raise ValueError(f"{column} column does not exist in {tbl.fullname} table")
 
-            for column in self.columns:
-                if not column in self.tbl.columns:
-                    raise ValueError(f"{column} column does not exist in {self.tbl.fullname} table")
+                self.delete_stmt = tbl.delete().where(
+                    sa.and_(
+                        *[(tbl.columns[column] == sa.bindparam(column)) for column in self.business_key_columns]))
 
-            self.delete_stmt = self.tbl.delete().where(
-                sa.and_(*[(self.tbl.columns[column] == sa.bindparam(column)) for column in self.business_key_columns]))
+                self.upsert_stmt = self.generate_upsert_stmt()
 
-            self.upsert_stmt = self.generate_upsert_stmt()
+        except OperationalError as e:
+            self.dispose_engine()
+            raise ConnectionError(e)
+
+    def dispose_engine(self):
+        with suppress(Exception):
+            self.connection.close()
+        with suppress(Exception):
+            self.engine.dispose()
+
+        self.business_key_columns = None
+        self.mapping_columns = None
+        self.columns = None
+        self.delete_stmt = None
+        self.upsert_stmt = None
+        self.tbl = None
+        self.connection = None
+        self.engine = None
```

Comment:
you can do this instead, IMO, so you don't need to specify them all and maintain it here when a new class member is added:
```
  for attr in vars(self):
        setattr(self, attr, None)
```

Reply:
✅